ecs-agent/tcs/handler/handler.go

```diff
@@ -0,0 +1,258 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+//lint:file-ignore U1000 Ignore unused metricsFactory field as it is only used by Fargate
+
+package tcshandler
+
+import (
+	"context"
+	"io"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/doctor"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/eventstream"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger/field"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/metrics"
+	tcsclient "github.com/aws/amazon-ecs-agent/ecs-agent/tcs/client"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/tcs/model/ecstcs"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/utils/retry"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/wsclient"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/cihub/seelog"
+)
+
+const (
+	// wsRWTimeout is the duration of read and write deadline for the websocket connection
+	deregisterContainerInstanceHandler = "TCSDeregisterContainerInstanceHandler"
+	ContainerRuntimeDocker             = "Docker"
+)
+
+// TelemetrySession defines an interface for handler's long-lived connection with TCS.
+type TelemetrySession interface {
+	StartTelemetrySession(context.Context, string) error
+	Start(context.Context) error
```

Comment:
Why do we need two similar functions?

Reply:
There are two reasons:
- First, we have backup and retry logic in `Start` that calls `StartTelemetrySession`, which does not have retry logic in it. Having these separated will improve the readability and make it easier for us to test their individual functionality, as well as providing embed option to override it in docker/containerd agent.
- Second, there is an extra string parameter to passed into `StartTelemetrySession`, this is the endpoint string. Currently, `DiscoverTelemetryEndpoint` requires `api.ecsClient`, which has a lot of dependencies, and some of them are not natural to move to `ecs-agent` module. As pointed in the first point, with this setup, we can have an overridden `Start` function to do retries based on combined `DiscoverTelemetryEndpoint` and `StartTelemetrySession` result and do not need to duplicate too much code.